src/csharp/Microsoft.Spark.Worker/Utils/AssemblyLoaderHelper.cs

```diff
@@ -0,0 +1,163 @@
+        ///
+        /// Note: Because <see cref="SparkContext.AddFile(string, bool)"/> prevents
+        /// overwriting/deleting files once they have been added to the
+        /// <see cref="SparkContext"/>, numbered identifiers are added to relevant files:
+        /// - <see cref="DependencyProviderUtils.FilePattern"/>
+        /// </summary>
+        /// <param name="stageId">The current Stage ID</param>
+        internal static void RegisterAssemblyHandler(int stageId)
+        {
+            if (!EnvironmentUtils.GetEnvironmentVariableAsBool("DOTNET_SPARK_REPL_MODE") ||
+                (stageId == s_stageId))
+            {
+                return;
+            }
+
+            // For a given stage, it is sufficient to instantiate one DependencyProvider.
+            // However, the Worker process may be reused between stages. New nuget dependencies
+            // may be introduced between stages and a new DependencyProvider will need to be
+            // created that can resolve them.
+            lock (s_lock)
+            {
+                if (stageId == s_stageId)
+                {
+                    return;
+                }
+                s_stageId = stageId;
+
+                string sparkFilesPath = SparkFiles.GetRootDirectory();
+                string metadataFile =
+                    FindHighestFile(sparkFilesPath, DependencyProviderUtils.FilePattern);
+
+                if (string.IsNullOrEmpty(metadataFile) || metadataFile.Equals(s_lastFileRead))
+                {
+                    return;
+                }
+                s_lastFileRead = metadataFile;
+
+                using FileStream fileStream = File.OpenRead(metadataFile);
+                BinaryFormatter formatter = new BinaryFormatter();
+                DependencyProviderUtils.Metadata metadata =
+                    (DependencyProviderUtils.Metadata)formatter.Deserialize(fileStream);
+
+                string unpackPath =
+                    Path.Combine(Directory.GetCurrentDirectory(), Path.Combine(".nuget", "packages"));
+                Directory.CreateDirectory(unpackPath);
+
+                IEnumerable<string> AssemblyProbingPaths()
+                {
+                    foreach (string dependency in metadata.AssemblyProbingPaths)
+                    {
+                        yield return Path.Combine(unpackPath, dependency);
+                    }
+                }
+
+                IEnumerable<string> NativeProbingRoots()
+                {
+                    foreach (string dependency in metadata.NativeProbingPaths)
+                    {
+                        yield return Path.Combine(unpackPath, dependency);
+                    }
+                }
+
+                UnpackPackages(sparkFilesPath, unpackPath, metadata.NuGets);
+
+                (s_dependencyProvider as IDisposable)?.Dispose();
+                s_dependencyProvider = new DependencyProvider(AssemblyProbingPaths, NativeProbingRoots);
```

Comment:
I guess this wouldn't affect the running tasks? Note that stages can run in parallel.

Reply:
I was under the impression that a Worker process would only be running parallel tasks for a specific stage.  Can a Worker process run multiple stages in parallel ?  If so, will we be locking too often ?

Changed the dispose order a bit.  I now register a new `DependencyProvider` for the new set of Assembly/Native probing paths before disposing the previously registered one.